src/sentry/event_manager.py

```diff
@@ -821,6 +828,23 @@ def _get_or_create_release_many(jobs, projects):
                     pop_tag(job["data"], "dist")
                     set_tag(job["data"], "sentry:dist", job["dist"].name)
 
+            with locks.get(
+                generate_dynamic_sampling_lock_key(project_id),
+                duration=60,
+                name="dynamic-sampling-lock",
+            ).acquire():
+                # Dynamic Sampling - Boosting latest release functionality
+                observed_release_in_last_day_key = generate_cache_key_for_observed_release(
+                    project_id, release.id
+                )
+                release_observed_in_last_24h = cache.get(observed_release_in_last_day_key)
+                cache.set(observed_release_in_last_day_key, 1, ONE_DAY_TIMEOUT)
```

Comment:
We should use redis' `SET` with the `GET` option (same as `GETSET` command) to do this atomically. With that, you can move the lock inside the branch below. 

Without this, you're acquiring a lock every time, which would come with a large performance penalty.